### ISSUE TITLE
Download CSV for all Inventory Items yields bad gateway error in csv file

### ISSUE THREAD

**Issue by joannekoch** [WHOIGit/ooicgsn-roundabout#168]
Search for all Inventory Items (2,326 items) and Download the CSV. The resulting downloaded file containing bad gateway is attached.

<img width="425" alt="Capture" src="https://user-images.githubusercontent.com/65416034/93094021-5ec51b00-f66f-11ea-8625-e4523cf21229.PNG">

**Comment by sbatchelder** [comment]
I just performed a full inventory CSV download on my local machine and in came through fine. This might be tricky to debug/fix if I can't reproduce the issue. 

It could be that the csv takes too long to generate and that the server is configured to respond with a 502 in that case. On my computer (2325 items) it does take a few minutes to compile the full csv and it would only get longer with more records. 

@joannekoch Is this issue consistent/reproducible on your end and would it be possible to see the server-side logs for the download request? Also, what are the `Timeout` and `KeepAliveTimeout` settings for the server you're using?

**Comment by joannekoch** [comment]
Hi Sidney,

Interesting…  It does take a while to compile the csv file – I timed it at 30 seconds. I can reproduce it every time on the Staging DB.

How do I get the logs and the timer settings from the Server?

Thanks,
Joanne

**Comment by sbatchelder** [comment]
@ethanandrews 
Do you know where the django logs on the server live? 
Also, any advice on this timeout issue? The ooi-rdb.conf configuration file under git:/compose/production/nginx says `proxy_send_timeout 90;`

It would be possible to increase this but that's not a good solution. 
Probably triggering a background job / thread on download would be a better solution, but I don't think that's a built-in django feature. It would mean a rework of the csv-search result downloading schema (and possibly other dynamically large downloads too).

**Comment by ethanandrews** [comment]
The logs for nginx are in the docker volume "production_nginx" -> /var/log/nginx

The Django error for this is a System Exit 1 error, which isn't very helpful and usuall just means that the process took too long. Full stack trace is available in Sentry: https://sentry.io/organizations/whoi/issues/1893884262/?project=277745&query=is%3Aunresolved

I think @sbatchelder is correct that this will require some reworking of the backend functionality. Some sort of asynchronous background job is probably required while the user's browser connection is kept alive (this is one of those things that you see dynamic progress bars for). Another possibility is caching the results so the DB is not queried every time, would probably work for a standard download like "download all" but not sure how valid that is with dynamic searches.

**Comment by sbatchelder** [comment]
I used [StreamingHttpResponse](https://docs.djangoproject.com/en/2.2/ref/request-response/#django.http.StreamingHttpResponse) to address the issue. The csv search results are streamed to the browser, keeping the request/connection alive, preventing the BadGateway error, and also improving download speed performance.

@MarioCarloni This may be relevant to your download needs. 
I found [THIS](https://github.com/django-import-export/django-import-export/issues/206#issuecomment-567935233) stackoverflow note helpful for me because my search results table uses tablib to export/download the data.
[HERE](https://stackoverflow.com/a/49820427) is a stackoverflow response that has a more generic approach. 
In both cases the key is to pass the data to export to `StreamingHttpResponse` via a generator object.
